0000-plugins.md

```diff
@@ -0,0 +1,155 @@
+	- [new_post_vote](https://github.com/LemmyNet/lemmy/blob/main/crates/db_schema/src/source/post.rs#L157)
+- Comment
+	- [new_comment](https://github.com/LemmyNet/lemmy/blob/main/crates/db_schema/src/source/comment.rs#L27)
+	- [new_comment_vote](https://github.com/LemmyNet/lemmy/blob/main/crates/db_schema/src/source/comment.rs#L105)
+
+## Example
+
+Below is a simple Go plugin which uses the `create_local_post` hook to replace `Rust` in post body with `Go`, and reject posts which mention `Java`. 
+
+Also checkout the documentation for Extism's [Go Plugin Development Kit](https://github.com/extism/go-pdk).
+
+```golang
+package main
+
+import (
+	"github.com/extism/go-pdk"
+	"strings"
+)
+
+type CreatePost struct {
+	Name string `json:"name"`
+	Body *string `json:"body,omitempty"`
+	...
+  }
+
+//export create_local_post
+func my_plugin() int32 {
+	params := CreatePost{}
+	// Read post data
+	err := pdk.InputJSON(&params)
+	if err != nil {
+		pdk.SetError(err)
+		return 1
+	}
+	if strings.Contains(params.Body, "Java") {
+		// Throw error to reject post
+		pdk.SetError("We dont talk about Java)
+		return 1
+	}
+	params.Body = strings.Replace(params.Body, "Rust", "Go", -1);
+	// Write back post data
+	err = pdk.OutputJSON(params)
+	if err != nil {
+		pdk.SetError(err)
+		return 1
+	}
+	return 0
+}
+
+func main() {}
+```
+
+## Host Functions
+
+Plugins will need access to the Lemmy database for various reasons, such as determining which instance a new post is coming from, or storing custom data. The easiest way to do this would be direct database access for plugins, but this has many problems:
+- Database schema changes across Lemmy versions, which would make it very difficult to make plugins compatible with new Lemmy versions, and prevent instances from upgrading
+- Plugins could accidentally delete data or break the database
+
+As such we need to use a different approach, by exposting specific database calls from Lemmy via [host functions](https://extism.org/docs/concepts/host-functions). For the beginning the following functions will be exposed:
+
+- Instance::read_or_create 
+
+More functions may be added if they are needed for the initial plugins.
+
+## Initial Plugins
+
+These can mainly serve as proof of concept, and as examples which can be modified by plugin developers. Include integration tests to ensure that functionality works with new Lemmy changes. These plugins should be written in a variety of languages. It could make sense to put them into a separate repo.
+
+- Block only downvotes from a specific instance
+- Replace words in post or comment text (eg "cloud to butt")
+- Replace Youtube URLs in post links with [Invidious](https://invidious.io/)
+- Push notifications (send data of new posts to localhost url, [#2631](https://github.com/LemmyNet/lemmy/issues/2631))
+- Language detection (automatically apply language tag to posts/comments, [#2870](https://github.com/LemmyNet/lemmy/issues/2870))
+- Additional suggestions welcome
```

Comment:
A few others I thought of:

- Forwarding links of new posts, comments, or likes to email, chat servers, or other platforms.
- Copying the previous version of a newly updated post or comment somewhere, to be quickly recalled.